File: app/src/main/java/com/privat/contacts/presentation/users/UsersAdapter.java
package com.privat.contacts.presentation.users;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.privat.contacts.databinding.UserItemBinding;
import com.privat.contacts.presentation.users.model.UserItemUi;

import java.util.ArrayList;
import java.util.List;
import java.util.function.Consumer;

public class UsersAdapter extends RecyclerView.Adapter<UsersAdapter.UserViewHolder> {
    private final ArrayList<UserItemUi> userItems = new ArrayList();
    private final Consumer<Integer> favoriteClick;
    private final Consumer<Integer> onClick;
    private final boolean showingFavorite;

    public UsersAdapter(Consumer<Integer> onClick) {
        this.onClick = onClick;
        showingFavorite = false;
        favoriteClick = (id) -> {
        };
    }

    public UsersAdapter(Consumer<Integer> onClick, Consumer<Integer> favoriteClick) {
        this.favoriteClick = favoriteClick;
        this.showingFavorite = true;
        this.onClick = onClick;
    }

    public void updateData(List<UserItemUi> newItems) {
        userItems.clear();
        userItems.addAll(newItems);
    }

    @NonNull
    @Override
    public UserViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        UserItemBinding binding = UserItemBinding.inflate(LayoutInflater.from(parent.getContext()), parent, false);
        return new UserViewHolder(binding);
    }

    @Override
    public void onBindViewHolder(@NonNull UserViewHolder holder, int position) {
        holder.bind(userItems.get(position));
    }

    @Override
    public int getItemCount() {
        return userItems.size();
    }

    class UserViewHolder extends RecyclerView.ViewHolder {
        private final UserItemBinding userItemBinding;

        UserViewHolder(@NonNull UserItemBinding userItemBinding) {
            super(userItemBinding.getRoot());
            this.userItemBinding = userItemBinding;
        }

        void bind(UserItemUi userItemUi) {
            userItemUi.showTitle(userItemBinding.tvName);
            userItemUi.showText(userItemBinding.tvPhone);
            if (showingFavorite)
                userItemUi.showFavorite(userItemBinding.ibFavorite, favoriteClick);
            else
                userItemBinding.ibFavorite.setVisibility(View.GONE);
        }
    }
}

File: app/src/main/java/com/privat/contacts/presentation/users/UsersAdapter.kt
package com.privat.contacts.presentation.users

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import com.privat.contacts.databinding.UserItemBinding
import com.privat.contacts.presentation.users.model.UserItemUi
import java.util.function.Consumer

class UsersAdapter : RecyclerView.Adapter<UsersAdapter.UserViewHolder> {
    private val userItems = ArrayList<UserItemUi>()
    private val deleteClick: Consumer<Int>
    private val onClick: Consumer<Int>
    private val showingFavorite: Boolean

    constructor(onClick: Consumer<Int>) {
        this.onClick = onClick
        showingFavorite = false
        deleteClick = Consumer { id: Int? -> }
    }

    constructor(onClick: Consumer<Int>, favoriteClick: Consumer<Int>) {
        deleteClick = favoriteClick
        showingFavorite = true
        this.onClick = onClick
    }

    fun updateData(newItems: List<UserItemUi>) {
        userItems.clear()
        userItems.addAll(newItems)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): UserViewHolder {
        val binding = UserItemBinding.inflate(LayoutInflater.from(parent.context), parent, false)
        return UserViewHolder(binding)
    }

    override fun onBindViewHolder(holder: UserViewHolder, position: Int) {
        holder.bind(userItems[position])
    }

    override fun getItemCount(): Int {
        return userItems.size
    }

    inner class UserViewHolder(private val userItemBinding: UserItemBinding) : RecyclerView.ViewHolder(userItemBinding.root) {
        fun bind(userItemUi: UserItemUi) {
            userItemUi.showTitle(userItemBinding.tvName)
            userItemUi.showText(userItemBinding.tvPhone)
            userItemUi.loadAvatar(userItemBinding.ivAvatar)
            if (showingFavorite) userItemUi.showFavorite(userItemBinding.ibFavorite, deleteClick) else userItemBinding.ibFavorite.visibility = View.GONE
            userItemBinding.root.setOnClickListener { listener: View? -> onClick.accept(userItemUi.id()) }
        }
    }
}
